protocols/autonatv2/src/server/behaviour.rs

```diff
@@ -0,0 +1,235 @@
+        Poll::Pending
+    }
+}
+
+impl<R> NetworkBehaviour for Behaviour<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    type ConnectionHandler = Handler<R>;
+
+    type ToSwarm = ();
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        _connection_id: ConnectionId,
+        _peer: PeerId,
+        _local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        Ok(Either::Right(dial_request::Handler::new(
+            remote_addr.clone(),
+            self.rng.clone(),
+        )))
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        _role_override: Endpoint,
+        port_use: PortUse,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if port_use == PortUse::New {
+            self.handlers.insert(
+                (
+                    addr.iter()
+                        .filter(|e| !matches!(e, Protocol::P2p(_)))
+                        .collect(),
+                    peer,
+                ),
+                connection_id,
+            );
+        }
+        Ok(Either::Left(dial_back::Handler::new()))
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            FromSwarm::DialFailure(DialFailure {
+                error: DialError::WrongPeerId { .. },
+                peer_id: Some(peer_id),
+                ..
+            }) => {
+                if let Some(key) = self
+                    .dialing_dial_back
+                    .keys()
+                    .find(|(_, p)| *p == peer_id)
+                    .cloned()
+                {
+                    let cmds = self.dialing_dial_back.remove(&key).unwrap();
+                    for cmd in cmds {
+                        let _ = cmd.back_channel.send(DialBack::Dial);
+                    }
+                }
+            }
+            FromSwarm::DialFailure(DialFailure {
+                error: DialError::Transport(pairs),
+                ..
+            }) => {
+                for (addr, _) in pairs.iter() {
+                    let cleaned_addr: Multiaddr = addr
+                        .iter()
+                        .filter(|p| !matches!(p, Protocol::P2p(_)))
+                        .collect();
+                    let peer_id_opt = addr.iter().find_map(|p| match p {
+                        Protocol::P2p(peer) => Some(peer),
+                        _ => None,
+                    });
+                    if let Some(peer_id) = peer_id_opt {
+                        if let Some(cmd) = self.dialing_dial_back.remove(&(cleaned_addr, peer_id)) {
+                            for cmd in cmd {
+                                let _ = cmd.back_channel.send(DialBack::Dial);
+                            }
+                        }
+                    }
```

Comment:
it would be easier to do:

```suggestion
			let Some(peer_id) = addr.iter().last() else {
				continue;
			};
                    
			if let Some(cmd) = self.dialing_dial_back.remove(&(cleaned_addr, peer_id)) {
			    for cmd in cmd {
			        let _ = cmd.back_channel.send(DialBack::Dial);
			    }
			}
```

Reply:
I think that's rendered unnecessary by the change to connection id.